### ISSUE TITLE
Compute Capability changes for CudaContext::computeThreadBlockSize

### ISSUE THREAD

**Issue by informatorius** [openmm/openmm#2901]
In CudaContext::computeThreadBlockSize the maxShared is hard coded. But it does not apply the values for compute capabilities > 3.0 so hard coded should be extended or values read from device
```
int CudaContext::computeThreadBlockSize(double memory, bool preferShared) const {
    int maxShared = 16*1024;
    if (computeCapability >= 2.0 && preferShared)
        maxShared = 48*1024;
    int max = (int) (maxShared/memory);
    if (max < 64)
        return 32;
    int threads = 64;
    while (threads+64 < max)
        threads += 64;
    return threads;
}
```
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#features-and-technical-specifications__technical-specifications-per-compute-capability

**Comment by informatorius** [comment]
Maybe also cudaOccupancyMaxPotentialBlockSize and similar since CUDA 6.5 could be helpful?
https://developer.nvidia.com/blog/cuda-pro-tip-occupancy-api-simplifies-launch-configuration/

**Comment by informatorius** [comment]
In CudaNonbondedUtilities forceThreadBlockSize is fixed to getComputeCapability() < 2.0 ? 128 : 256
Is this the best value on all hardware?
```
CudaNonbondedUtilities::CudaNonbondedUtilities(CudaContext& context) : context(context), useCutoff(false), usePeriodic(false), anyExclusions(false), usePadding(true),
        blockSorter(NULL), pinnedCountBuffer(NULL), forceRebuildNeighborList(true), lastCutoff(0.0), groupFlags(0), canUsePairList(true) {
    // Decide how many thread blocks to use.

    string errorMessage = "Error initializing nonbonded utilities";
    int multiprocessors;
    CHECK_RESULT(cuDeviceGetAttribute(&multiprocessors, CU_DEVICE_ATTRIBUTE_MULTIPROCESSOR_COUNT, context.getDevice()));
    CHECK_RESULT(cuEventCreate(&downloadCountEvent, 0));
    CHECK_RESULT(cuMemHostAlloc((void**) &pinnedCountBuffer, 2*sizeof(int), CU_MEMHOSTALLOC_PORTABLE));
    numForceThreadBlocks = 4*multiprocessors;
    forceThreadBlockSize = (context.getComputeCapability() < 2.0 ? 128 : 256);
    setKernelSource(CudaKernelSources::nonbonded);
}
```

**Comment by peastman** [comment]
`computeThreadBlockSize()` is a mostly obsolete method that's only ever used in AMOEBA, and will hopefully go away altogether when I rewrite the AMOEBA code (see #2700).  Don't worry about it.

> Is this the best value on all hardware?

So far as I know, yes.  It's one of the things I test when each new generation comes out, and so far that value has always worked well.
